Make ActiveStorage::Service responsible for checksums

mrpasquini opened the pull request:
### Motivation / Background

This PR moves responsibility for providing the class used to calculate checksums to the ActiveStorage::Service class. This makes the storage service class fully responsible for calculating digests on the back end and removes the need to expose the digest class or ActiveStorage.checksum_implementation in other classes.

### Detail
S3, GCS, and Azure all support file integrity checksums in addition to MD5. MD5 is not the recommended algorithm by most of these cloud services and cannot be used in a FIPS compliant environment. This is a first step toward allowing these service classes to support additional/alternative algorithms for file integrity.

This PR:
- removes the private implementation of `calculate_checksum_in_chunks` from Blob in favor a new public method `calculate_checksum` on the storage service class to manage all digest calculation for that service. 
- ~removes direct use of `ActiveStorage.checksum_implementation` by other class/services when verifying files in favor of `calculate_checksum`.~ Removes direct use of OpenSSL::Digest::MD5 by services in favor of a new public `checksum_implementation` method on the base service class. This method returning `OpenSSL::Digest::MD5` by default to maintain backwards compatibility.
- removes the separate ActiveStorage::Downloader class in favor of a combined private `download_and_verify_tempfile` implementation in the base service.
- refactors direct uploads to separate algorithms from framework functionality.

This PR also implements SHA256 support for the S3 service including support for direct uploads.


See #54123, #34443 or #54040 for additional discussion. 


### Additional information

<!-- Provide additional information such as benchmarks, references to other repositories, or alternative solutions. -->

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

byroot commented on activestorage/lib/active_storage/downloader.rb:
```diff
@@ -35,7 +35,7 @@ def download(key, file)
 
       def verify_integrity_of(file, checksum:)
-        unless ActiveStorage.checksum_implementation.file(file).base64digest == checksum
+        unless service.file(file) == checksum
```
`file` doesn't look like a good name. Perhaps `file_checksum` or something?

mrpasquini replied:
`file_base64digest` or `base64digest_file` may be more clear. It is adding a `base64digest` call onto the digest class's `file` method.

byroot commented on activestorage/lib/active_storage/service.rb:
```diff
@@ -148,6 +148,31 @@ def public?
+      checksum_implementation.base64digest(io)
+    end
+
+    def file(file)
```
Or even `base64digest_path`

byroot reviewed (commented):
Thank you, this is indeed way more actionable.

I would like to see this new API used by one service though, whichever you think is simpler. Once the new API is proven to work for one, we can convert the other ones in independent PRs.

seanpdoyle commented on activestorage/app/models/active_storage/blob.rb:
```diff
@@ -246,7 +246,7 @@ def upload(io, identify: true)
 
   def unfurl(io, identify: true) # :nodoc:
-    self.checksum     = compute_checksum_in_chunks(io)
+    self.checksum     = service&.compute_checksum_in_chunks(io)
```
Similar to https://github.com/rails/rails/pull/54468/files#r1947500486, I wonder how we might improve the `compute_checksum_in_chunks` naming. 

Since it was previously a `private` method, it's likely that its name served as its only documentation. If it were promoted to the public API, it could be more thoroughly documented and wouldn't need to communicate argument type expectations in its name.

What do you think about omitting the `_in_chunks` suffix?

```suggestion
    self.checksum     = service&.compute_checksum(io)
```

mrpasquini replied:
Would it be clear what the difference would be between `base64digest` vs `compute_checksum`?

We could keep `compute_checksum_in_chunks` private on the base service class and expand `base64digest` to something like
`def base64digest(io, in_chunks: false)`

byroot replied:
I think it would be simpler if we assume `checksum_implementation` respond to `new` and `<<`.

mrpasquini replied:
I looking to avoid having to call `.base64digest` on the return of `file` or `<<` (in compute_checksum_in_chunks/compute_checksum method) but it would be possible to remove `base64digest` and `file` public methods on the services if we wanted to change calls from:
`service.file(file)` to `service.checksum_implementation.file(file).base64digest)`
and
`service.base64digest(io)` to `service.checksum_implementation.base64digest(io)`

mrpasquini commented:
> Thank you, this is indeed way more actionable.
> 
> I would like to see this new API used by one service though, whichever you think is simpler. Once the new API is proven to work for one, we can convert the other ones in independent PRs.

Can you clarify what you mean by one service using this new API? All services are in this PR inherit and using the methods from the base service class. There are not many places that call these methods directly.

The next unit of work is to make the algorithm explicit instead of implicitly MD5 (roughly 300 line change). I see two main choices:
(A) adding algorithm as a separate kwarg/param, etc similar to what is here:
https://github.com/rails/rails/commit/e131557de02e949f960762c17aebef9c9383fdc4#diff-50ea5d4b0963e7dee40b48eee0875e0287578ece2c2360f7e0d3897df0d26d09
vs
(B) Similar to how it was implemented in #54123 as a simple class with digest and algorithm attributes.

I am currently leaning toward A because it seems like a lower impact on the services API. We have to add kwargs support to `url_for_direct_upload` and then the algorithm as a kwarg across the service, which seems a much lower impact than changing checksum to a class. Originally I had thought B may be a smaller, more simple fix, but they seemed to come out similar in what had to be touched. We could include the digest and algorithm storage to ActiveStorage::Blob.checksum either here or as a separate PR if this is too large a chunk of work.

The last parts would be the additional algorithms for each service. The commits in #54123 have this split out per service and are pretty good examples of what it would look like.

mrpasquini commented:
> > Can you clarify what you mean by one service using this new API?
> 
> By that I mean changing S3 or CGS to start using something else than MD5 to see how it looks and if there is some shortcomings with the API.
> 
> Also how it behave when one of these service start storing checksums with a prefix.

I really appreciate the feedback and quick response. I am concerned that adding functionality to have the services use something other than MD5 and storing with the prefix is going to add over 500+ lines to this PR and then we are in the same state as #54123.

The main purpose of this PR is removing direct calls to `ActiveStorage.checksum_implementation` and routing usage through the methods on the service class/adapters instead.

As far as the implementation of this, I have added 2 methods for convenience and moved one private method off AS:Blob to public on service. It would be possible to just use `checksum_implementation` on the service if we wanted to remove any of these 3 methods:
- `base64digest` method on service is just for ease of use. `ActiveStorage.checksum_implementation.base64digest` could be  `service.checksum_implementation.base64digest` instead of `service.base64digest`.
- `file` method on service is just for ease of use. `ActiveStorage.checksum_implementation.file(file).digest` to `service.checksum_implementation.file(file).digest` instead of `service.file(file)`. 
- `compute_checksum_in_chunks` was private on AS::Blob, but moved to the service to consolidate responsibility for digest calculations to the services. In the future, services could allow tweaking of this calculation process such differing chunk size rather than have it opaquely in AS:Blob. That said, if we wanted to leave it as a private method on AS:Blob, we do so easily by having it call `service.checksum_implementation` instead of `ActiveStorage.checksum_implementation`.



The A or B options from my last comment sound like what you are most interested in. I would prefer to open it as a follow on PR to this one as that is something like a 300 line PR own its own.

Adding support for storing non-MD5 to AS:Blob can be done independently. I can add a PR for that in parallel now if that is the best way to address it.

I dont see as much value in adding algorithm support to the services until AS can use them. The changes to support are some service specific methods to address differences in interacting with the SDK and direct HTTP Headers between MD5 and other algorithms. Adding support is going to be very similar to something like https://github.com/rails/rails/pull/54123/commits/a75186ce1c641082bb187fa1a0bff37a12d32163

byroot commented:
> I am concerned that adding functionality to have the services use something other than MD5 and storing with the prefix is going to add over 500+ lines to this PR and then we are in the same state as #54123.

Not sure if it was clear, but the idea was that:

  - Only do one backend (e.g. only S3)
  - Don't introduce a `Checksum` class, just have the backend know how to recognize the checksum.

I don't think that would be 500 lines, would it? Also just feel free to just to it in a branch on your fork, what I'd like to know if whether this refactoring is useful or if we're going to hit a wall and have made lots of changes for nothing.

mrpasquini commented:
> > I am concerned that adding functionality to have the services use something other than MD5 and storing with the prefix is going to add over 500+ lines to this PR and then we are in the same state as #54123.
> 
> Not sure if it was clear, but the idea was that:
> 
> * Only do one backend (e.g. only S3)
> * Don't introduce a `Checksum` class, just have the backend know how to recognize the checksum.
> 
> I don't think that would be 500 lines, would it? Also just feel free to just to it in a branch on your fork, what I'd like to know if whether this refactoring is useful or if we're going to hit a wall and have made lots of changes for nothing.

If we skip the encoder for ActiveStorage::Blob, then its about 150 lines, though I am probably missing tests, to add working support for S3. Its been added to this PR.

byroot commented on activestorage/CHANGELOG.md:
```diff
@@ -1,3 +1,12 @@
+*   Support additional file checksum algorithms in S3 Adapter
+
+    Add support for CRC32, CRC32c, SHA1, SHA256, and CRC64NVMe to S3 service for
```
I think that's where we might as misunderstood each others.

I don't think there is any value in supporting 6 different checksum algorithms for S3, the goal is just to change the current MD5 for something else (and to support both concurrently to allow smooth migration).

So we shouldn't add these 5, just one of these.

mrpasquini replied:
I would choose SHA256 for my original intention of making things easier for FIPS compliance. This is a different choice than raw throughput/performance, so it seemed to make sense to support multiple for the different use cases. There is not much different in the adapter itself beyond MD5 vs non-MD5.

The other thing to perhaps take into account is that multipart uploads do not currently support checksums in the s3 adapter. If this was to be added, there are some differences based on the checksum algorithm (full file vs composite checksum) as noted under "Multipart uploads" section [here](https://docs.aws.amazon.com/AmazonS3/latest/userguide/checking-object-integrity.html)

bogdan commented:
I wonder how if this code is somehow related to the improvement:

https://github.com/rails/rails/blob/main/activestorage/app/javascript/activestorage/file_checksum.js#L5

I didn't use it myself, but I thought it also needs to be updated, no?

mrpasquini commented:
> I wonder how if this code is somehow related to the improvement:
> 
> https://github.com/rails/rails/blob/main/activestorage/app/javascript/activestorage/file_checksum.js#L5
> 
> I didn't use it myself, but I thought it also needs to be updated, no?

As of now in this PR, direct upload cases are still using MD5. To support additional algorithms for direct upload cases we would need to pass the algorithm to use to the client, calculate the digest with the correct algorithm on the client, do the direct upload with correct headers for the cloud provider for the ActiveStorage::Service + algorithm, and have the form pass checksum to the direct upload controller as either `<md5digest>` or `<algo>:<digest>` string formats.

bogdan commented on activestorage/lib/active_storage/service.rb:
```diff
@@ -148,6 +148,56 @@ def public?
+      when :MD5
+        ActiveStorage.checksum_implementation
+      else
+        send(algorithm.downcase)
```
If I understood it correctly, I suppose to implement a method that matches the digest algoritm set. I don't understand this convention:

* I am not able to use custom algorithm with built-in service as it would throw no method error.
* if only MD5 is supported out of the box, I would need to subclass `ActiveStorage::Service` class to implement anything else. In this case why do I need to configure algorithm via yaml for my custom service?
* If I set `SHA-256` as algorithm, I am not sure which method I should implement. It generates additional ambiguity.

The use case I have in mind is IPFS Service implementation where file can only be retried via checksum and the checksum on top of digest has addtional codec information, so the length of IPFS hash is 36 bytes not 32.

I would be more happy to impelement `ActiveStorage::Services::IpfsService#compute_checksum_in_chunks` for that more than anything else because it is interface is clear unlike `checksum_implementation` which doesn't have defined return format.

mrpasquini replied:
Thank you for identifying the gap here allow a method missing. It was missed in the evolution to this branch. I understand your concerns to be that what should be service specific implementations have bled into the base service class. I agree.

I will update the 2nd commit in this PR to not update the base service class. The base service shouldn't be concerned with `default_digest_algorithm` or `SUPPORTED_CHECKSUM_ALGORITHMS` etc... Those would be S3 service specific implementation.

mrpasquini commented on activestorage/lib/active_storage/service.rb:
```diff
@@ -148,6 +148,56 @@ def public?
+    def checksum_implementation(algorithm = default_digest_algorithm)
+      case algorithm
+      when :MD5
+        ActiveStorage.checksum_implementation
```
The `ActiveStorage.checksum_implementation` is already merged and not part of this PR. Removing it seems like it would be a separate issue and separate PR. I do not have a problem removing it provided there is SHA256 support, but it I don't want to end up recreating an ever-growing PR like https://github.com/rails/rails/pull/54123 again.

bogdan replied:
I think the way to make it consistent and not overgrow is to remove `default_digest_algorithm` configuration option from this PR and reconsider it together with removing `checksum_implementation`.

mrpasquini replied:
`ActiveStorage.checksum_implementation` only supports alternative implementations of MD5, not additional algorithms. It could be renamed or dropped. I am happy to remove it as part of this PR and likely move fallback capability to the default service class like https://github.com/rails/rails/pull/54123/files#diff-ac8beaa42f6abaff6c92c76bba0340ac4784e81f8ab1fbc67a923dfc6d03d634R172-R190

bogdan commented on activestorage/lib/active_storage/errors.rb:
```diff
@@ -26,4 +26,7 @@ class FileNotFoundError < Error; end
   class PreviewError < Error; end
+
+  # Raised when a checksum is not supported by a service
+  class UnsupportedChecksumError < Error; end
```
Conventionally it should be named `ChecksumUnsupportedError` to follow general convesion "what? kind?" like "RecordNotFound", "InputInvalid" etc.

mrpasquini replied:
Removing as this is likely not needed until the direct upload implementation.

mrpasquini committed:
Make ActiveStorage::Service responsible for checksums

Expose new public methods in storage service and remove private
 implementation in Blob.

bogdan commented on activestorage/lib/active_storage/service.rb:
```diff
@@ -152,6 +152,23 @@ def inspect # :nodoc:
       "#<#{self.class}#{name.present? ? " name=#{name.inspect}" : ""}>"
     end
 
+    def compute_checksum(io, **)
```
Did I understand it correctly that custom storage impelmentations suppose to override this method?

In case of IPFS, it is absolutelly necessary because its hashing algorithm is using 256KB chunk and [Merkle DAG](https://docs.ipfs.tech/concepts/merkle-dag/) for chunk checksums (reading this doc might be too dangerous for unprepared person 😄 ).

mrpasquini replied:
Yes. We could move everything behind this method and no longer expose checksum_implementation or ActiveStorage.checksum_implementation outside of it. I think that would be ideal.

I initially thought this may add more complexity than I wanted to introduce, but we should be able to just check if `io.is_a?(File)` and return `checksum_implementation.file(io).base64digest` from compute_checksum if we continued to want to use the .file implementation.

My assumption about chunking is that the digest classes like `OpenSSL::Digest::MD5.base64digest(io)` do a full read from IO, which would not preform well if it if it did not fit in memory.

I can add a commit to enhance `compute_checksum` to use .file for file classes and remove use of `service.checksum_implementation.file(file).base64digest` in favor of `service.compute_checksum`

bogdan commented on activestorage/lib/active_storage/service.rb:
```diff
@@ -152,6 +152,23 @@ def inspect # :nodoc:
+
+      checksum_implementation.new.tap do |checksum|
+        read_buffer = "".b
+        while io.read(5.megabytes, read_buffer)
```
I researched the chunking topic. All sharing algorithms (md5, sha-2/3 etc) are shaving o(n) complexity, so chunking does give any performance benefit if not performed in parallel.

I am not sure why chunking appeared in the first place, but it will not give any performance benefit in the current form.

mrpasquini replied:
Rails has already been doing 5.megabyte chunks or deferring to the `file` implementation of the Digest class. I also do not think it would make sense to remove it.

I added a `default_chunk_size` configuration to ActiveStorage for the existing value (and added a case to disable it with `.to_i == 0`). Service classes could respect this or implement their own, perhaps tuned to the algorithm in use.

For new vs old data, the chunk size does not effect the calculation of the checksum value for MD5 so persisted data/checksums would continue to function with a changed configuration setting. Other algorithms may vary, but that would be the responsibility of the service classes that implement them.

mrpasquini commented on activestorage/lib/active_storage/service/disk_service.rb:
```diff
@@ -161,7 +161,7 @@ def make_path_for(key)
 
       def ensure_integrity_of(key, checksum)
-        unless ActiveStorage.checksum_implementation.file(path_for(key)).base64digest == checksum
+        unless checksum_implementation.file(path_for(key)).base64digest == checksum
```
If we do a File.open on this path, instead of passing the path, I think we can handle easier in `compute_checksum` as a file.

bogdan commented on activestorage/lib/active_storage/downloader.rb:
```diff
@@ -35,7 +35,7 @@ def download(key, file)
 
       def verify_integrity_of(file, checksum:)
-        unless ActiveStorage.checksum_implementation.file(file).base64digest == checksum
+        unless service.compute_checksum(file) == checksum
           raise ActiveStorage::IntegrityError
```
Needs to be a little more clear when raising exception:

```suggestion
        actual_checksum = service.compute_checksum(file) 
        unless actual_checksum == checksum
          raise ActiveStorage::IntegrityError, "Checksum verification failed expecting #{checksum}, but downloaded file having #{actual_checksum}"
```

bogdan commented on activestorage/lib/active_storage/downloader.rb:
```diff
@@ -35,8 +35,9 @@ def download(key, file)
       def verify_integrity_of(file, checksum:)
-        unless ActiveStorage.checksum_implementation.file(file).base64digest == checksum
-          raise ActiveStorage::IntegrityError
+        actual_checksum = service.compute_checksum(file)
```
This utility file is weird to have: it is only used in one place and only has one public method. How about merging it directly into service to make sure it is consistently overwritten if needed?

mrpasquini commented:
> @mrpasquini any plans on which version of Rails will have it?
I had lost track there was an open ask to add this to single service. Let me add something to the S3 service and see if we can get this to an acceptable place.

mrpasquini commented on activestorage/app/javascript/activestorage/file_checksum.js:
```diff
@@ -1,45 +1,64 @@
-      this.callback(null, base64digest)
+      const digest = this.algorithmConfig.finalize(this.checksumBuffer)
+      const formattedChecksum = this.algorithmConfig.formatChecksum(digest)
+      this.callback(null, formattedChecksum)
```
One did encoding, one did formatting. I will combine to a single method when I look at the test failures later this week.
